crates/libs/implement/src/gen.rs

```diff
@@ -0,0 +1,579 @@
+            // Foo_Impl to safe Rust code, because doing so would allow unsound behavior in safe Rust
+            // code, due to the adjustments of the reference count that Foo_Impl permits.
+            //
+            // This is why this function returns ComObject<Self> instead of returning #impl_ident.
+
+            fn into_object(self) -> ::windows_core::ComObject<Self> {
+                let boxed = ::windows_core::imp::Box::<#impl_ident::#generics>::new(self.into_outer());
+                unsafe {
+                    let ptr = ::windows_core::imp::Box::into_raw(boxed);
+                    ::windows_core::ComObject::from_raw(
+                        ::core::ptr::NonNull::new_unchecked(ptr)
+                    )
+                }
+            }
+        }
+    }
+}
+
+/// Generates the `query_interface` method.
+fn gen_query_interface(inputs: &ImplementInputs) -> syn::ImplItemFn {
+    let queries = inputs.interface_chains.iter().map(|interface_chain| {
+        let chain_ty = interface_chain.implement.to_vtbl_ident();
+        let chain_field = &interface_chain.field_ident;
+        quote_spanned! {
+            interface_chain.implement.span =>
+            if #chain_ty::matches(&iid) {
+                break 'found &self.#chain_field as *const _ as *const ::core::ffi::c_void;
+            }
+        }
+    });
+
+    // Dynamic casting requires that the object not contain non-static lifetimes.
+    let enable_dyn_casting = inputs.original_type.generics.lifetimes().count() == 0;
+    let dynamic_cast_query = if enable_dyn_casting {
+        quote! {
+            if iid == ::windows_core::DYNAMIC_CAST_IID {
+                // DYNAMIC_CAST_IID is special. We _do not_ increase the reference count for this pseudo-interface.
+                // Also, instead of returning an interface pointer, we simply write the `&dyn Any` directly to the
+                // 'interface' pointer. Since the size of `&dyn Any` is 2 pointers, not one, the caller must be
+                // prepared for this. This is not a normal QueryInterface call.
+                //
+                // See the `Interface::cast_to_any` method, which is the only caller that should use DYNAMIC_CAST_ID.
+                (interface as *mut *const dyn core::any::Any).write(self as &dyn ::core::any::Any as *const dyn ::core::any::Any);
+                return ::windows_core::HRESULT(0);
+            }
+        }
+    } else {
+        quote!()
+    };
+
+    let identity_query = quote! {
+        if iid == <::windows_core::IUnknown as ::windows_core::Interface>::IID
+        || iid == <::windows_core::IInspectable as ::windows_core::Interface>::IID
+        || iid == <::windows_core::imp::IAgileObject as ::windows_core::Interface>::IID {
+            break 'found &self.identity as *const _ as *const ::core::ffi::c_void;
+        }
+    };
+
+    let tear_off_query = quote! {
+        let tear_off_ptr = self.count.query(&iid, &self.identity as *const _ as *mut _);
```

Comment:
Just checking, we don't need an explicit `AddRef` here?

Reply:
This is the same as the previous code, just with slightly different control flow. I don't understand the tear-off interface stuff, so I'll defer to Kenny on that.

Reply:
The weak ref tearoff shares a ref count with the object. This is all handled by `WeakRefCount`.